### ISSUE TITLE
OpenXR in Godot 4

### ISSUE THREAD

**Issue by Darkheron** [godotengine/godot#42409]
**Godot version:**
4.0

**OS/device including version:**

Linux version 5.4.0-42-generic (buildd@lgw01-amd64-038) (gcc version 9.3.0 (Ubuntu 9.3.0-10ubuntu2))
Nvidia GeForce RGX 2080Ti
Vulkan 1.2
AMD Ryzen Threadripper 2950X 16-Core Processor

**Issue description:**
Hello. I am currently working on an implementation of OpenXR using the Vulkan backend that I have directly inserted into the Godot source. It is all going really well, but I have reached an impasse.  I don't understand what the vision is for integrating OpenXR with the viewports/windows and I need some guidance to understand the intent.  There is a SubViewport class and node that can be instantiated within the scene root that can be flagged to "Use XR". Is the intent to use these SubViewport classes somehow to feed the active XrInterface? I thought maybe so, but the viewport_to_screen_rect inside of the underlying RenderServerViewport::Viewport is always [0,0] because the viewport is never attached to the screen. So I clearly don't understand what is going on here.

Basically, my idea for getting this implementation working was to slowly undef the section of code in rendering_server_viewport.cpp between #if 0 and the endif (lines 369 to 410) and try to get the eye commits working; however, the SubViewport is always skipped in the render loop because `vp->last_pass != draw_viewports_pass` is always true because of line 334, ie.. `bool visible = vp->viewport_to_screen_rect != Rect2();`

Help.

**Steps to reproduce:**
Compile.. then run. :)

Is this even the correct place to ask this question? I kinda seems like a bug but I dunno.
TIA

**Comment by fire** [comment]
Are you able to show the current code as a PR to Godot Engine? We can look at your code and comment easier regardless of @BastiaanOlij 's status of merging.

**Comment by Darkheron** [comment]
I guess I can try. It is forked onto gitlab and it is pretty nasty as I am just learning OpenXR and how it integrates with Vulkan.  It will definitely undergo a rewrite. 

I not sure it would even be helpful though because what I really wish to know is how the Godot project is supposed to be set up so that xr_interface->commit_for_eye can even get called.  At this point my SubViewport is skipped within the RenderingServerViewport::draw_viewports method because it is never attached to the screen. What I really need to know is if using a SubViewport is even what I should be doing in 4.0 or if I am way off base in how an XR viewport is to be established within a Godot session.  Maybe I am doing a bad job of explaining. :)

**Comment by BastiaanOlij** [comment]
Note that all OpenXR development at this point is focused on Godot 3.2, there is no support for Vulkan in the XRInterface yet. We'll need to do some major changes to how the external texture support is implemented and how indeed commit_for_eye will work. 

You can find the plugin here: https://github.com/godotvr/godot_openxr

Obviously any ideas or progress you've made on getting Vulkan to work with XR would be greatly appreciated if shared. I've been somewhat busy but I really want to sink my teeth into this sooner rather then later.

**Comment by Darkheron** [comment]
The more I look into this the more I kinda think the whole approach might be wrong for 4.0. I mean is there even a need for an XRServer and generic XRInterface given that OpenXR is system agnostic in the first place?  I am starting to lean towards the approach used in the OpenXR hello_xr code where a concrete XRInterface class uses specific graphics API plugins for GL and Vulkan. But I might be missing something.

Many questions remain.

I might just blow this whole thing up to try something new and see what happens.

**Comment by BastiaanOlij** [comment]
@Darkheron , unfortunately the industry isn't at that point yet. While I am hopeful the industry will adopt OpenXR that adoption has only just begun with SteamVR and Oculus still having their OpenXR solution in beta and many other vendors, especially those in the AR space seeing many of those run on Android and there is no OpenXR support in Android itself.

So the XRServer will remain critical in Godot 4 even if OpenXR will likely turn out to be our main focus for AR and VR support. What will happen after that and whether we'll have tighter integration with OpenXR if the future makes that a viable choice, I think that is definately a good scenario for the future.

So yeah, agree with your line of thought here, but timing wise the industry isn't ready yet.

Finally keep in mind that the vast majority of developers are still developing non-XR games and to them having XR build into the core is just dead weight. This is another reason why we've chosen to implement all XR solutions as plugins leaving the bare minimum in the core.

**Comment by KoBeWi** [comment]
Is this issue still relevant?

**Comment by fire** [comment]
As far as I know openxr is in core godot engine. So this can be closed as completed
